nixos/modules/services/misc/huginn.nix

```diff
@@ -0,0 +1,728 @@
+        };
+	minPassword = mkOption {
+          type = types.int;
+	  default = 8;
+	  description = "Minimum length for user passwords";
+	};
+	resetWithin = mkOption {
+	  type = types.str;
+	  default = "6.hours";
+	  description = "Duration for which reset password token is valid";
+	};
+	enableUserLockout = mkOption {
+	  type = types.bool;
+	  default = true;
+	  description = "If true lock user accounts for the unlockAfter period they fail maxFailedLogin login attempts. If false to allow unlimited failed login attempts.";
+	};
+	maxFailedLogin = mkOption {
+	  type = types.int;
+	  default = 10;
+	  description = "After how many failed login attempts the account is unlocked when lockStrategy is set to failed_attempts.";
+	};
+	unlockStrategy = mkOption {
+	  type = types.str;
+	  default = "both";
+	  description = "Can be set to 'email', 'time', 'both' or 'none'. 'none' requires manual unlocking of your users!";
+	};
+	unlockAfter = mkOption {
+	  type = types.str;
+	  default = "1.hour";
+	  description = "Duration after which the user is unlocked.";
+	};
+	sessionDuration = mkOption {
+	  type = types.str;
+	  default = "4.weeks";
+	  description = "Duration for which the user will be remembered without asking for creditials again.";
+	};
+	enableScenarioDefaults = mkOption {
+	  type = types.bool;
+	  default = true;
+	  description = "Set to 'true' if you would prefer new users to start with a default set of agents";
+	};
+	defaultScenario = mkOption {
+	  type = types.str;
+	  default = "";
+	  description = "Path to a different default scenario for users to use. If blank uses standard default scenario.";
+	};
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+
+    environment.systemPackages = [ huginn-rake ];
+
+    # We use postgres as the main data store.
+    services.postgresql.enable = mkDefault true;
+
+    users.extraUsers = [
+      { name = cfg.user;
+        group = cfg.group;
+        home = "${cfg.statePath}/";
+        shell = "${pkgs.bash}/bin/bash";
+        uid = config.ids.uids.huginn;
+      }
+    ];
+
+    users.extraGroups = [
+      { name = cfg.group;
+        gid = config.ids.gids.huginn;
+      }
+    ];
+
+    systemd.services.huginn = {
+      after = [ "network.target" "postgresql.service" ];
+      wantedBy = [ "multi-user.target" ];
+      environment = filteredHuginnEnv;
+      path = with pkgs; [
+        config.services.postgresql.package
+      ];
+      preStart = ''
+        rm -rf ${cfg.statePath}/tmp
+        mkdir -p ${cfg.statePath}/log
+        mkdir -p ${cfg.statePath}/tmp/pids
+	mkdir -p ${cfg.statePath}/tmp/cache
+        mkdir -p ${cfg.statePath}/tmp/sockets
+
+        touch ${cfg.statePath}/tmp/secrets
+	if [ "${cfg.appSecretToken}" = "" ]; then
+	  printf "APP_SECRET_TOKEN=" >>  ${cfg.statePath}/tmp/secrets
+          tr -dc A-Za-z0-9 < /dev/urandom 2> /dev/null | head -c 32 >> ${cfg.statePath}/tmp/secrets
+	  echo "" >> ${cfg.statePath}/tmp/secrets
```

Comment:
maybe this could be also wrapped and reused in a shell function

```bash
gen_secret() {
  printf "$1=%s\n" "$(tr -dc A-Za-z0-9 < /dev/urandom 2> /dev/null|head -c32)" > "$2"
}

gen_secret "APP_SECRET_TOKEN" "${cfg.statePath}/tmp/secrets"
```